### PR TITLE
Add option for AWS S3 bucket installation filestores

### PR DESCRIPTION
**Pull request by gabrieljackson**
This change includes many enhancements to allow the cloud server
to provision and manage the AWS resources necessary for a secure
S3 bucket installation filestore. To do this, the AWS IAM, S3,
and Secrets Manager services are used.

Primary changes:
 - Change database schema to store filestore type for installations.
 - Use IAM service to create users, policies, and access keys for
   each installation requiring an S3 filestore.
 - Use S3 for the filestore bucket.
 - Use Secrets Manager to store IAM access keys in a secure way.
 - Provide a flag for server startup that will optionally allow
   S3 filestore teardown to preserve the S3 bucket data as an
   extra precaution.
 - All generic filestore operations have been moved to interfaces.

Other changes:
 - Change database schema to store database type for installations.
   This lays the foundation for AWS RDS installation databases.
 - New installation states for pre-provisioning and cleanup. This
   allows for better state handling when an error is encountered.
 - Cluster Installation creation failure will now retry in most
   cases when a failure is encountered.
 - Bug fix for setting the wrong deletion failure state in one
   case.

https://mattermost.atlassian.net/browse/MM-19081

### CODE REVIEW

**Comment by cpanato** [comment]
will take some time to read :)

File: cmd/cloud/server.go [hunk 1]
```diff
@@ -40,6 +40,7 @@ func init() {
 	serverCmd.PersistentFlags().String("public-subnets", "", "The public subnet IDs to use on AWS.")
 	serverCmd.PersistentFlags().Int("poll", 30, "The interval in seconds to poll for background work.")
 	serverCmd.PersistentFlags().Int("cluster-resource-threshold", 80, "The percent threshold where new installations won't be scheduled on a multi-tenant cluster")
+	serverCmd.PersistentFlags().Bool("keep-filestore-data", false, "Whether to preserve filestore data after installation deletion or not")
```
Comment (jwilander) [inline cmd/cloud/server.go]: Should we default to true for safety?

File: internal/tools/aws/aim.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
+package aws
```
Comment (jwilander) [inline internal/tools/aws/aim.go]: Typo in the filename? "aim.go" vs "iam.go"?

File: model/installation_filestore.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+	mmv1alpha1 "github.com/mattermost/mattermost-operator/pkg/apis/mattermost/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// InstallationFilestoreMinioOperator is a filestore hosted in kubernetes via the operator.
+	InstallationFilestoreMinioOperator = "minio-operator"
+	// InstallationFilestoreAwsS3 is a filestore hosted via Amazon S3.
+	InstallationFilestoreAwsS3 = "aws-s3"
```
Comment (jwilander) [inline model/installation_filestore.go]:
```suggestion
	InstallationFilestoreAWSS3 = "aws-s3"
```

Should we capitalize AWS?

Comment (gabrieljackson) [reply model/installation_filestore.go]: I did initially, but found it a bit hard to read so wasn't sure. I am fine with either version.

File: internal/tools/aws/aim.go (hunk 2 above)
Comment (gabrieljackson) [reply internal/tools/aws/aim.go]: Yep, thanks!

File: model/installation_filestore.go (hunk 3 above)
Comment (jwilander) [reply model/installation_filestore.go]: 0/5 just saw us capitalizing other acronyms in other places

File: cmd/cloud/server.go (hunk 1 above)
Comment (gabrieljackson) [reply cmd/cloud/server.go]: Yeah, I wasn't sure if we wanted to err on the side of cleaning up by default or not and having to do it manually. Either one is fine with me.

Comment (jwilander) [reply cmd/cloud/server.go]: I'd lean towards not cleaning up by default

**Comment by cpanato** [comment]
thanks, this is very cool

**Pull request merged**